core/src/main/java/io/grpc/PickFirstBalancerFactory2.java

```diff
@@ -92,30 +115,24 @@ public void handleNameResolutionError(Status error) {
 
     @Override
     public void handleSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
-      if (subchannel != this.subchannel || stateInfo.getState() == ConnectivityState.SHUTDOWN) {
-        return;
-      }
+      ConnectivityState currentState = stateInfo.getState();
+      AtomicReference<ConnectivityState> lastStateRef = subchannel.getAttributes().get(LAST_STATE);
+      checkState(lastStateRef != null, "subchannel should have LAST_STATE attribute set");
 
       PickResult pickResult;
-      switch (stateInfo.getState()) {
-        case READY:
-        case IDLE:
-          pickResult = PickResult.withSubchannel(subchannel);
-          break;
-        case TRANSIENT_FAILURE:
-          pickResult = PickResult.withError(stateInfo.getStatus());
-          break;
-        case CONNECTING:
-        case SHUTDOWN:
-          pickResult = PickResult.withNoResult();
-          break;
-        default:
-          throw new IllegalStateException();
+      if (subchannel != this.subchannel || currentState == SHUTDOWN || currentState == CONNECTING) {
```

Comment:
@zhangkun83 should we throw on `IllegallStateException` on `subchannel == null` or is it fine to just return no result?